tests/unit/type/zset.tcl

```diff
@@ -1025,6 +1089,112 @@ start_server {tags {"zset"}} {
     basics listpack
     basics skiplist
 
+    test "ZPOP/ZMPOP against wrong type" {
+        r set foo{t} bar
+        assert_error "*WRONGTYPE*" {r zpopmin foo{t}}
+        assert_error "*WRONGTYPE*" {r zpopmax foo{t}}
+        assert_error "*WRONGTYPE*" {r zpopmin foo{t} 2}
+
+        assert_error "*WRONGTYPE*" {r zmpop 1 foo{t} min}
+        assert_error "*WRONGTYPE*" {r zmpop 1 foo{t} max}
+        assert_error "*WRONGTYPE*" {r zmpop 1 foo{t} max count 200}
+
+        r del foo{t}
+        r set foo2{t} bar
+        assert_error "*WRONGTYPE*" {r zmpop 2 foo{t} foo2{t} min}
+        assert_error "*WRONGTYPE*" {r zmpop 2 foo2{t} foo1{t} max count 1}
+    }
+
+    test "ZMPOP with illegal argument" {
+        assert_error "ERR wrong number of arguments*" {r zmpop}
+        assert_error "ERR wrong number of arguments*" {r zmpop 1}
+        assert_error "ERR wrong number of arguments*" {r zmpop 1 myzset{t}}
+
+        assert_error "ERR numkeys*" {r zmpop 0 myzset{t} MIN}
+        assert_error "ERR numkeys*" {r zmpop a myzset{t} MIN}
+        assert_error "ERR numkeys*" {r zmpop -1 myzset{t} MAX}
+
+        assert_error "ERR syntax error*" {r zmpop 1 myzset{t} bad_where}
+        assert_error "ERR syntax error*" {r zmpop 1 myzset{t} MIN bar_arg}
+        assert_error "ERR syntax error*" {r zmpop 1 myzset{t} MAX MIN}
+        assert_error "ERR syntax error*" {r zmpop 1 myzset{t} COUNT}
+        assert_error "ERR syntax error*" {r zmpop 2 myzset{t} myzset2{t} bad_arg}
+
+        assert_error "ERR count*" {r zmpop 1 myzset{t} MIN COUNT 0}
+        assert_error "ERR count*" {r zmpop 1 myzset{t} MAX COUNT a}
+        assert_error "ERR count*" {r zmpop 1 myzset{t} MIN COUNT -1}
+        assert_error "ERR count*" {r zmpop 2 myzset{t} myzset2{t} MAX COUNT -1}
+    }
+
+    test "ZMPOP against non existing key - nil" {
+        r del zset{t} zset2{t}
+
+        assert_equal {} [r zmpop 1 zset{t} min]
+        assert_equal {} [r zmpop 1 zset{t} max count 1]
+        assert_equal {} [r zmpop 2 zset{t} zset2{t} min]
+        assert_equal {} [r zmpop 2 zset{t} zset2{t} max count 1]
+
+        # Make sure we will reply a NIL.
+        r readraw 1
+        assert_equal {*-1} [r zmpop 1 zset{t} min]
+        assert_equal {*-1} [r zmpop 1 zset{t} max count 1]
+        assert_equal {*-1} [r zmpop 2 zset{t} zset2{t} min]
+        assert_equal {*-1} [r zmpop 2 zset{t} zset2{t} max count 1]
+        r readraw 0
+
+        r hello 3
+        r readraw 1
+        assert_equal {_} [r zmpop 1 zset{t} min]
+        assert_equal {_} [r zmpop 1 zset{t} max count 1]
+        assert_equal {_} [r zmpop 2 zset{t} zset2{t} min]
+        assert_equal {_} [r zmpop 2 zset{t} zset2{t} max count 1]
+        r readraw 0
+        r hello 2
```

Comment:
let's include a similar test with readraw for the other variants and code paths.
i.e. ZPOPMIN, BZPOPMIN, BZPOP.
in the blocking variants, let's verify both the blocking paths and the non-blocking path.
or maybe there's another place where this will be easier to cover?

Reply:
I camp up with this. (also bunch of code...)
i have to say, test code is much harder to write than expected
In most cases, i have to write a lot of code...